In [1]:
import pandas as pd
import math
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [2]:
df = pd.read_csv(r'C:\Users\samas\Downloads\NYU 2021 returns.csv')

In [3]:
df1 = pd.read_csv(r'C:\Users\samas\Downloads\Final 2021 value signal.csv')

## Q1

To compute the annualised daily average excess returns, I first computed the average daily excess returns for the dataset and annualised them. If I first annualise the daily returns and compute the average, due to the presence of some outliers, the resulting statistic is affected producing incredible numbers. This is the reason I first computed the average and then annualised the results. 

In [4]:
excess_rets = pd.DataFrame(columns= ['excess_5Y', 'excess_tips','excess_nasdaq','excess_unemp'])

In [5]:
excess_rets['excess_5Y'] = df['5Y'] - df['1M']
excess_rets['excess_tips'] = df['5Y TIPS'] - df['1M']
excess_rets['excess_nasdaq'] = df['5Y Nasdaq'] - df['1M']
excess_rets['excess_unemp'] = df['5Y Unemp'] - df['1M']

In [6]:
print("\033[1mAnnualised Average Excess Returns")
annualised_means = (1 + excess_rets.mean())**365 -1
annualised_means

Annualised Average Excess Returns


excess_5Y        0.025511
excess_tips      0.004473
excess_nasdaq    0.112189
excess_unemp     0.023199
dtype: float64

In [7]:
print('\033[1mAnnualised Standard Deviation of Excess Returns')
annualised_std = (excess_rets.std())*sqrt(365)
annualised_std

Annualised Standard Deviation of Excess Returns


excess_5Y        0.050740
excess_tips      0.060025
excess_nasdaq    0.098448
excess_unemp     0.041873
dtype: float64

In [8]:
print('\033[1mAnnualised Sharpe Ratios')
annualised_means/annualised_std

Annualised Sharpe Ratios


excess_5Y        0.502776
excess_tips      0.074523
excess_nasdaq    1.139570
excess_unemp     0.554040
dtype: float64

## Q2

To compute CAGR, I first assumed that that the price is 1 initially, and grew according to the excess returns from 2003 to 2021. And calculated the CAGRs for all securities using that dataframe.

In [9]:
id_5y_excess = [1]
id_tips_excess = [1]
id_nasdaq_excess = [1]
id_unemp_excess = [1]
for index, row in excess_rets.iterrows():
    val = (row['excess_5Y'] +1) * id_5y_excess[-1]
    val1 = (row['excess_tips'] + 1) * id_tips_excess[-1]
    val2 = (row['excess_nasdaq'] + 1) * id_nasdaq_excess[-1]
    val3 = (row['excess_unemp'] + 1) * id_unemp_excess[-1]
    id_5y_excess.append(val)
    id_tips_excess.append(val1)
    id_nasdaq_excess.append(val2)
    id_unemp_excess.append(val3)

In [10]:
indices_excess = pd.DataFrame(list(zip(id_5y_excess,id_tips_excess,id_nasdaq_excess,id_unemp_excess)))
indices_excess.columns = ['5Y','tips','nasdaq','unemp']

In [11]:
indices_excess

5Y      tips    nasdaq     unemp
0     1.000000  1.000000  1.000000  1.000000
1     1.000973  1.000016  0.996712  1.001395
2     0.997903  0.998158  0.993371  1.004163
3     1.000713  0.999603  0.990072  1.004779
4     1.002148  1.003439  0.997535  1.002405
...        ...       ...       ...       ...
4581  1.352026  1.029755  3.543196  1.328037
4582  1.350796  1.030157  3.549493  1.326549
4583  1.348849  1.030650  3.561912  1.324632
4584  1.350198  1.033303  3.573286  1.321310
4585  1.350230  1.034338  3.579916  1.319248

[4586 rows x 4 columns]

In [12]:
print("\033[1mCompound Average Annual Excess Growth Rates")
(indices_excess.iloc[-1]/indices_excess.iloc[1])**(1/18.25) - 1

Compound Average Annual Excess Growth Rates


5Y        0.016535
tips      0.001851
nasdaq    0.072575
unemp     0.015220
dtype: float64

In [13]:
print("\033[1mAnnualised Average Excess Returns")
annualised_means

Annualised Average Excess Returns


excess_5Y        0.025511
excess_tips      0.004473
excess_nasdaq    0.112189
excess_unemp     0.023199
dtype: float64

## The difference between the Annualised Average excess returns and CAGR is plausible. The computation is different for both the measures, and CAGR takes into account only the beginning value and ending value and then computes the average growth thoughout the period. But Average excess returns is average of all the returns over the period which says us the average return from that security, whereas CAGR tells us the average growth in the value of security per year over the period. They basically have different interpretations and different methods of computation.

## Q3

In [14]:
print("\033[1mCovariance Matrix")
excess_rets.cov()

Covariance Matrix


excess_5Y  excess_tips  excess_nasdaq  excess_unemp
excess_5Y       0.000007     0.000005       0.000003     -0.000004
excess_tips     0.000005     0.000010       0.000002     -0.000004
excess_nasdaq   0.000003     0.000002       0.000027     -0.000008
excess_unemp   -0.000004    -0.000004      -0.000008      0.000005

In [15]:
print('\033[1mCorrelation Matrix')
excess_rets.corr()

Correlation Matrix


excess_5Y  excess_tips  excess_nasdaq  excess_unemp
excess_5Y       1.000000     0.648624       0.248629     -0.665537
excess_tips     0.648624     1.000000       0.133158     -0.622251
excess_nasdaq   0.248629     0.133158       1.000000     -0.746471
excess_unemp   -0.665537    -0.622251      -0.746471      1.000000

In [16]:
print("\033[1mInverse Covariance Matrix")
pd.DataFrame(np.linalg.inv(excess_rets.cov()), excess_rets.columns, columns= ['excess_5Y','excess_tips','excess_nasdaq','excess_unemp'])

Inverse Covariance Matrix


excess_5Y    excess_tips  excess_nasdaq  excess_unemp
excess_5Y      344583.518688  -38943.919020   81773.495706  3.866752e+05
excess_tips    -38943.919020  282051.402721  117435.691521  4.262890e+05
excess_nasdaq   81773.495706  117435.691521  162255.335496  4.554673e+05
excess_unemp   386675.224866  426289.039002  455467.302633  1.699638e+06

## Both the correlation and covariance between 5Y Unemp and 5Y TIPS is negative which supports the Philips Curve result of negative relationship between inflation and unemployment. Since the securities 5Y Unemp and 5Y TIPS are like hedge against unemployment and inflation respectively, the performance of the securities is based on unemployment and inflation levels at that point of time and the negative correlation is consistent with the theory of the Philips Curve.

## Q4

In [17]:
sharpe_weights = np.linalg.inv(excess_rets.cov()).dot(excess_rets.mean()) # computing the optimal weights

In [18]:
sharpe_weights = sharpe_weights/sum(sharpe_weights)

In [19]:
print('\033[1mOptimal Sharpe Weights')
sharpe_weights

Optimal Sharpe Weights


array([0.1465    , 0.12667589, 0.17017907, 0.55664504])

In [20]:
mean_excess_returns_sharpe = sharpe_weights.dot(excess_rets.mean())
mean_excess_returns_sharpe

9.621989329719596e-05

In [21]:
standard_deviation_excess_rets_sharpe = sqrt(sharpe_weights.dot(excess_rets.cov()).dot(sharpe_weights))
standard_deviation_excess_rets_sharpe

0.000444238248663218

In [22]:
print('\033[1mSharpe ratio of Optimal Sharpe Portfolio ')
portfolio_sharpe = mean_excess_returns_sharpe/standard_deviation_excess_rets_sharpe
portfolio_sharpe

Sharpe ratio of Optimal Sharpe Portfolio 


0.21659524722766801

In [23]:
annualised_mean_excess_rets_sharpe = (1 + mean_excess_returns_sharpe)**365 -1
annualised_stdev_excess_rets_sharpe = standard_deviation_excess_rets_sharpe * sqrt(365)
annualised_portfolio_sharpe = annualised_mean_excess_rets_sharpe/annualised_stdev_excess_rets_sharpe
print('\033[1mAnnualised Sharpe ratio of Optimal Sharpe Portfolio')
annualised_portfolio_sharpe

Annualised Sharpe ratio of Optimal Sharpe Portfolio


4.21136302757986

In [24]:
annualised_mean_excess_rets_sharpe

0.03574251109118265

In [25]:
annualised_stdev_excess_rets_sharpe

0.008487159823816653

## Q5

## Computing Returns of three factor portfolios

In [26]:
df['mkt-rf'] = 0.25*df['5Y'] + 0.25*df['5Y TIPS'] + 0.25*df['5Y Nasdaq'] + 0.25*df['5Y Unemp'] - df['1M']

In [27]:
momentum = []
for index, row in df.loc[1:].iterrows():
    weight_5y = df['5Y'][index-1] * 10
    weight_tips = df['5Y TIPS'][index-1] * 10
    weight_nasdaq = df['5Y Nasdaq'][index-1] * 10
    weight_unemp = df['5Y Unemp'][index-1] * 10
    total = weight_5y + weight_tips + weight_nasdaq + weight_unemp
    weight_1m = -1 * total
    rets = weight_5y * row['5Y'] + weight_tips * row['5Y TIPS'] + weight_nasdaq * row['5Y Nasdaq'] + weight_unemp * row['5Y Unemp'] + weight_1m * row['1M']
    momentum.append(rets)

In [28]:
momentum_pf = pd.DataFrame(columns = ['DATE','Returns'])

In [29]:
momentum_pf['DATE'] = df['DATE']

In [30]:
momentum_pf.drop(index = 0, inplace = True)

In [31]:
momentum_pf['Returns'] = momentum

In [32]:
df1['1M'] = -1 * (df1['5Y'] + df1['5Y TIPS'] + df1['5Y Nasdaq'] + df1['5Y Unemp'])

In [33]:
value_rets = []
for index, row in df.iterrows():
    rets = df1['5Y'][index+1] * row['5Y'] + df1['5Y TIPS'][index+1] * row['5Y TIPS'] + df1['5Y Nasdaq'][index+1] * row['5Y Nasdaq'] + df1['5Y Unemp'][index+1] * row['5Y Unemp'] + df1['1M'][index+1] * row['1M']
    value_rets.append(rets)

In [34]:
df['value'] = value_rets

In [35]:
factor_pfs = pd.DataFrame(columns = ['mkt-rf','momentum','value'])

In [36]:
factor_pfs['mkt-rf'] = df['mkt-rf'][1:]
factor_pfs['value'] = df['value'][1:]
factor_pfs['momentum'] = momentum

In [37]:
print('\033[1mReturns from Three Factor Portfolios')
factor_pfs

Returns from Three Factor Portfolios


mkt-rf  momentum     value
1    -0.001378  0.000117  0.000555
2     0.000389  0.000017  0.000627
3     0.002612 -0.000166 -0.001409
4    -0.003526 -0.000781  0.000380
5     0.000039 -0.000036  0.000662
...        ...       ...       ...
4580 -0.000058  0.000131 -0.000008
4581  0.000034  0.000051  0.000002
4582  0.000273  0.000093 -0.000003
4583  0.001065  0.000146 -0.000010
4584  0.000330  0.000124 -0.000024

[4584 rows x 3 columns]

In [38]:
factor_pfs['rf'] = df['1M'][1:]

In [39]:
anuualised_avg_fps_rets = (1 + factor_pfs.mean())**365 - 1
print('\033[1mAnnualised Average Returns of Factor Portfolios')
anuualised_avg_fps_rets[:3]

Annualised Average Returns of Factor Portfolios


mkt-rf      0.040559
momentum    0.038295
value       0.001788
dtype: float64

In [40]:
annualised_std_fps_rets = (factor_pfs.std())*sqrt(365)
print('\033[1mAnnualised Standard Deviations of returns from Factor Portfolios')
annualised_std_fps_rets[:3]

Annualised Standard Deviations of returns from Factor Portfolios


mkt-rf      0.029073
momentum    0.009511
value       0.002691
dtype: float64

In [41]:
sharpe_mkt_rf = ((1 + ((factor_pfs['mkt-rf'] - factor_pfs['rf']).mean()))**365 - 1)/(((factor_pfs['mkt-rf'] - factor_pfs['rf']).std())* sqrt(365))
sharpe_momentum = ((1 + ((factor_pfs['momentum'] - factor_pfs['rf']).mean()))**365 - 1)/(((factor_pfs['momentum'] - factor_pfs['rf']).std())* sqrt(365))
sharpe_value = ((1 + ((factor_pfs['value'] - factor_pfs['rf']).mean()))**365 - 1)/(((factor_pfs['value'] - factor_pfs['rf']).std())* sqrt(365))

In [42]:
print('\033[1mAnnualised Sharpe ratio of Mkt-Rf portfolio')
sharpe_mkt_rf

Annualised Sharpe ratio of Mkt-Rf portfolio


0.7757616342824938

In [43]:
print('\033[1mAnnualised Sharpe ratio of Momentum portfolio')
sharpe_momentum

Annualised Sharpe ratio of Momentum portfolio


2.1907977137950065

In [44]:
print('\033[1mAnnualised Sharpe ratio of Value portfolio')
sharpe_value

Annualised Sharpe ratio of Value portfolio


-4.9766195821458155

## Q6

## The annualised Sharpe ratio of the Optimal Sharpe portfolio is 4.21136302757986 as computed in Q4. And none of the ratios of three factor portfolios is greater than 4.21136302757986. It is not possible for the factor portfolios computed above to have a higher Sharpe ratio in Q4 because, the optimization in part 4 is done in such a way that maximizes the Sharpe Ratio of the portfolio and thus the weights are derived, so its not possible for the three factor portfolios to have a better sharpe than the one in Q4, but the mean returns are higher for the factor portfolios and the standard deviations are also high with an exception to value portfolio compared to the optimal sharpe portfolio. So we get an optimal sharpe ratio from Q4 but we can create portfolios with higher mean returns with higher risk using different strategies and since the optimisation is done with the objective of maximizing the sharpe ratio in part 4 we get those weights where the sharpe is maximized and those might not produce greater mean returns necessarily. That is the trade off for using optimal sharpe weights.

## Q7

In [45]:
import statsmodels.api as sm

In [46]:
Y = factor_pfs['mkt-rf']
X = factor_pfs[['momentum','value']]
X= sm.add_constant(X)
model = sm.OLS(Y,X)
model.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 mkt-rf   R-squared:                       0.004
Model:                            OLS   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     9.604
Date:                Wed, 12 May 2021   Prob (F-statistic):           6.89e-05
Time:                        18:48:45   Log-Likelihood:                 23246.
No. Observations:                4584   AIC:                        -4.649e+04
Df Residuals:                    4581   BIC:                        -4.647e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       8.844e-05   2.29e-05      3.859      0.000    4.35e-05       0.000
momentum       0.1943      0.045      4.308      0.000       0.106       0.283
value          0.0976      0.159      0.612      0.540      -0.215       0.410
==============================================================================
Omnibus:                     3388.442   Durbin-Watson:                   1.807
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           260110.690
Skew:                           2.866   Prob(JB):                         0.00
Kurtosis:                      39.455   Cond. No.                     7.11e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.11e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [47]:
Y1 = factor_pfs['momentum']
X1 = factor_pfs[['mkt-rf','value']]
X1= sm.add_constant(X1)
model1 = sm.OLS(Y1,X1)
model1.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               momentum   R-squared:                       0.006
Model:                            OLS   Adj. R-squared:                  0.006
Method:                 Least Squares   F-statistic:                     13.90
Date:                Wed, 12 May 2021   Prob (F-statistic):           9.58e-07
Time:                        18:48:45   Log-Likelihood:                 28372.
No. Observations:                4584   AIC:                        -5.674e+04
Df Residuals:                    4581   BIC:                        -5.672e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       9.994e-05   7.36e-06     13.587      0.000    8.55e-05       0.000
mkt-rf         0.0208      0.005      4.308      0.000       0.011       0.030
value          0.1556      0.052      2.989      0.003       0.054       0.258
==============================================================================
Omnibus:                     7544.119   Durbin-Watson:                   1.742
Prob(Omnibus):                  0.000   Jarque-Bera (JB):         13324310.887
Skew:                          10.496   Prob(JB):                         0.00
Kurtosis:                     266.287   Cond. No.                     7.10e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 7.1e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [48]:
Y2 = factor_pfs['value']
X2 = factor_pfs[['momentum','mkt-rf']]
X2= sm.add_constant(X2)
model2 = sm.OLS(Y2,X2)
model2.fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  value   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     4.791
Date:                Wed, 12 May 2021   Prob (F-statistic):            0.00835
Time:                        18:48:45   Log-Likelihood:                 34151.
No. Observations:                4584   AIC:                        -6.830e+04
Df Residuals:                    4581   BIC:                        -6.828e+04
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.517e-06   2.13e-06      1.654      0.098   -6.51e-07    7.68e-06
momentum       0.0125      0.004      2.989      0.003       0.004       0.021
mkt-rf         0.0008      0.001      0.612      0.540      -0.002       0.004
==============================================================================
Omnibus:                     4419.521   Durbin-Watson:                   2.054
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          4585322.511
Skew:                           3.594   Prob(JB):                         0.00
Kurtosis:                     157.775   Cond. No.                     2.01e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.01e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [49]:
arrays = [['Mkt-Rf','Mkt-Rf','Momentum','Momentum','Value','Value'],['coeff','t-stat','coeff','t-stat','coeff','t-stat']]

In [50]:
regressions = pd.DataFrame(columns = ['Alpha','Mkt-Rf','Momentum','Value','R-squared'],index= arrays)

In [51]:
regressions['Alpha'] = [8.844e-05,3.859,9.994e-05,13.587,3.517e-06,1.654]
regressions['Mkt-Rf'] = [NaN,NaN,0.0208,4.308,0.0008,0.612]
regressions['Momentum'] = [0.1943,4.308,NaN,NaN,0.0125,2.989]
regressions['Value'] = [0.0976,0.612,0.1556,2.989,NaN,NaN]
regressions['R-squared'] = [0.004,NaN,0.006,NaN,0.002,NaN]

In [52]:
print('\033[1mRegressions Table')
regressions

Regressions Table


Alpha  Mkt-Rf  Momentum   Value  R-squared
Mkt-Rf   coeff    0.000088     NaN    0.1943  0.0976      0.004
         t-stat   3.859000     NaN    4.3080  0.6120        NaN
Momentum coeff    0.000100  0.0208       NaN  0.1556      0.006
         t-stat  13.587000  4.3080       NaN  2.9890        NaN
Value    coeff    0.000004  0.0008    0.0125     NaN      0.002
         t-stat   1.654000  0.6120    2.9890     NaN        NaN

## The results indicate a positive association among all the three portfolios. From the first regression, there is a significant and positive relationship between Mkt-Rf and Momentum portfolios but  not with the value portfolio. In the second regression, we see that the relationship is significant and positive for both Mkt-Rf and Value against Momentum portfolio. In the third regression, we see a significance of relationship between the value and momentum but not the Mkt-Rf. And all the R-squared are low as expected. The Alphas are significant for all the regressions. From the regressions we can say that no strong association exists between Value and Mkt-Rf portfolios.


## Q8

In [53]:
from scipy.stats import norm

In [54]:
def bsm_call(sigma, S, K, r, t): #Black Scholes Formula
    d1 = (log(S/K) + (r + sigma**2/2)* t)/(sigma * sqrt(t))
    d2 = d1- sigma*sqrt(t)
    c = norm.cdf(d1) *S - norm.cdf(d2) * K * math.exp(-r * t)
    return c

## Creating the Indices from the returns

In [55]:
id_5y = [1]
id_tips = [1]
id_nasdaq = [1]
id_unemp = [1]
for index, row in df.iterrows():
    val = (row['5Y'] +1) * id_5y[-1]
    val1 = (row['5Y TIPS'] + 1) * id_tips[-1]
    val2 = (row['5Y Nasdaq'] + 1) * id_nasdaq[-1]
    val3 = (row['5Y Unemp'] + 1) * id_unemp[-1]
    id_5y.append(val)
    id_tips.append(val1)
    id_nasdaq.append(val2)
    id_unemp.append(val3)

In [56]:
indices = pd.DataFrame(list(zip(id_5y,id_tips,id_nasdaq,id_unemp)))

In [57]:
indices.columns = ['5Y','tips','nasdaq','unemp']

In [58]:
indices

5Y      tips    nasdaq     unemp
0     1.000000  1.000000  1.000000  1.000000
1     1.001005  1.000048  0.996744  1.001427
2     0.998032  0.998288  0.993499  1.004292
3     1.000875  0.999764  0.990232  1.004941
4     1.002342  1.003633  0.997728  1.002599
...        ...       ...       ...       ...
4581  1.677056  1.277361  4.398113  1.646897
4582  1.675532  1.277860  4.405932  1.645052
4583  1.673118  1.278472  4.421350  1.642676
4584  1.674792  1.281763  4.435470  1.638557
4585  1.674832  1.283048  4.443701  1.636001

[4586 rows x 4 columns]

## To compute the Call Price, I am using the past 21 days historical standard deviation of returns as the sigma in BSM. Here is the computation of the Standard deviations series

In [59]:
import statistics as st
std_5y = []
std_tips = []
std_nasdaq = []
std_unemp = []
for index, row in df.loc[21:].iterrows():
    t = []
    tips = []
    nasdaq = []
    unemp = []
    for i in range(index-21,index):
        a = df['5Y'][i]
        t.append(a)
        b = df['5Y TIPS'][i]
        tips.append(b)
        c = df['5Y Nasdaq'][i]
        nasdaq.append(c)
        d = df['5Y Unemp'][i]
        unemp.append(d)
 
    std1 = st.stdev(t)
    std_5y.append(std1)
    std2 = st.stdev(tips)
    std_tips.append(std2)
    std3 = st.stdev(nasdaq)
    std_nasdaq.append(std3)
    std4 = st.stdev(unemp)
    std_unemp.append(std4)

In [60]:
date = ['1/2/2003']
dates = list(df['DATE'])
dates = date + dates

In [61]:
indices['DATE'] = dates

In [62]:
indices.index -= 1

In [63]:
indices['1M'] = df['1M']

In [64]:
indices = indices.iloc[22:]

In [65]:
indices['5y_std'] = std_5y
indices['tips_std'] = std_tips
indices['nasdaq_std'] = std_nasdaq
indices['unemp_std'] = std_unemp

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in

In [66]:
indices

5Y      tips    nasdaq     unemp       DATE            1M  \
21    1.005496  1.015970  1.022678  1.007957   2/4/2003  3.178080e-05   
22    1.001876  1.015523  1.021515  1.009985   2/5/2003  3.232880e-05   
23    1.004269  1.018496  1.018519  1.010112   2/6/2003  3.232880e-05   
24    1.006669  1.020007  1.023230  1.008665   2/7/2003  3.205480e-05   
25    1.004132  1.020122  1.020342  1.010650  2/10/2003  9.616440e-05   
...        ...       ...       ...       ...        ...           ...   
4580  1.677056  1.277361  4.398113  1.646897  4/23/2021  5.479450e-07   
4581  1.675532  1.277860  4.405932  1.645052  4/26/2021  8.219180e-07   
4582  1.673118  1.278472  4.421350  1.642676  4/27/2021  5.479450e-07   
4583  1.674792  1.281763  4.435470  1.638557  4/28/2021  2.739730e-07   
4584  1.674832  1.283048  4.443701  1.636001  4/29/2021  2.739730e-07   

        5y_std  tips_std  nasdaq_std  unemp_std  
21    0.003062  0.002524    0.005882   0.002349  
22    0.003115  0.002621    0.005832   0.002391  
23    0.003150  0.002573    0.005769   0.002352  
24    0.003132  0.002611    0.005755   0.002351  
25    0.003157  0.002525    0.005630   0.002309  
...        ...       ...         ...        ...  
4580  0.001630  0.002132    0.001969   0.000818  
4581  0.001642  0.002112    0.001972   0.000826  
4582  0.001623  0.002117    0.001976   0.000822  
4583  0.001599  0.002047    0.001932   0.000811  
4584  0.001608  0.002046    0.001928   0.000809  

[4564 rows x 10 columns]

## Computing the Call Prices
I am taking time t as 21 since all data points are on daily basis.

In [67]:
call_5y = []
call_tips = []
call_nasdaq = []
call_unemp = []
for index, row in indices.iterrows():
    y5 = bsm_call(row['5y_std'],row['5Y'],row['5Y'],row['1M'],21)
    call_5y.append(y5)
    tips = bsm_call(row['tips_std'],row['tips'],row['tips'],row['1M'],21)
    call_tips.append(tips)
    nasdaq = bsm_call(row['nasdaq_std'],row['nasdaq'],row['nasdaq'],row['1M'],21)
    call_nasdaq.append(nasdaq)
    unemp = bsm_call(row['unemp_std'],row['unemp'],row['unemp'],row['1M'],21)
    call_unemp.append(unemp)

In [68]:
indices['call_5y'] = call_5y
indices['call_tips'] = call_tips
indices['call_nasdaq'] = call_nasdaq
indices['call_unemp'] = call_unemp

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in

In [69]:
indices

5Y      tips    nasdaq     unemp       DATE            1M  \
21    1.005496  1.015970  1.022678  1.007957   2/4/2003  3.178080e-05   
22    1.001876  1.015523  1.021515  1.009985   2/5/2003  3.232880e-05   
23    1.004269  1.018496  1.018519  1.010112   2/6/2003  3.232880e-05   
24    1.006669  1.020007  1.023230  1.008665   2/7/2003  3.205480e-05   
25    1.004132  1.020122  1.020342  1.010650  2/10/2003  9.616440e-05   
...        ...       ...       ...       ...        ...           ...   
4580  1.677056  1.277361  4.398113  1.646897  4/23/2021  5.479450e-07   
4581  1.675532  1.277860  4.405932  1.645052  4/26/2021  8.219180e-07   
4582  1.673118  1.278472  4.421350  1.642676  4/27/2021  5.479450e-07   
4583  1.674792  1.281763  4.435470  1.638557  4/28/2021  2.739730e-07   
4584  1.674832  1.283048  4.443701  1.636001  4/29/2021  2.739730e-07   

        5y_std  tips_std  nasdaq_std  unemp_std   call_5y  call_tips  \
21    0.003062  0.002524    0.005882   0.002349  0.005969   0.005033   
22    0.003115  0.002621    0.005832   0.002391  0.006049   0.005217   
23    0.003150  0.002573    0.005769   0.002352  0.006128   0.005143   
24    0.003132  0.002611    0.005755   0.002351  0.006107   0.005218   
25    0.003157  0.002525    0.005630   0.002309  0.006859   0.005805   
...        ...       ...         ...        ...       ...        ...   
4580  0.001630  0.002132    0.001969   0.000818  0.005009   0.004985   
4581  0.001642  0.002112    0.001972   0.000826  0.005046   0.004945   
4582  0.001623  0.002117    0.001976   0.000822  0.004973   0.004956   
4583  0.001599  0.002047    0.001932   0.000811  0.004902   0.004801   
4584  0.001608  0.002046    0.001928   0.000809  0.004927   0.004804   

      call_nasdaq  call_unemp  
21       0.011338    0.004671  
22       0.011238    0.004764  
23       0.011087    0.004694  
24       0.011109    0.004682  
25       0.011553    0.005360  
...           ...         ...  
4580     0.015857    0.002473  
4581     0.015923    0.002499  
4582     0.015994    0.002477  
4583     0.015680    0.002433  
4584     0.015674    0.002424  

[4564 rows x 14 columns]

## Computing Discounted Values at expiry as maximum of 0 or difference between the Index value at Expiry and the Strike Price of Call and discounted at 1M rate at the expiry of call. 
## I cannot compute the discounted value at expiry for the last 21 days, so I am excluding those.

In [70]:
indices.index -= 21
value_5y = []
value_tips = []
value_nasdaq = []
value_unemp = []
for index, row in indices.loc[21:].iterrows():
    val = max(row['5Y'] - indices['5Y'][index-21], 0)
    val = math.exp(-1 * row['1M'] * 21)*val
    val1 = max(row['tips'] - indices['tips'][index-21], 0)
    val1 = math.exp(-1 * row['1M'] * 21)*val1
    val2 = max(row['nasdaq'] - indices['nasdaq'][index-21], 0)
    val2 = math.exp(-1 * row['1M'] * 21)*val2
    val3 = max(row['unemp'] - indices['unemp'][index-21], 0)
    val3 = math.exp(-1 * row['1M'] * 21)*val3
    value_5y.append(val)
    value_tips.append(val1)
    value_nasdaq.append(val2)
    value_unemp.append(val3)

In [71]:
disc_values_at_exp = pd.DataFrame(list(zip(value_5y,value_tips,value_nasdaq,value_unemp)),columns=['5Y','tips','nasdaq','unemp'])

In [72]:
disc_values_at_exp

5Y      tips    nasdaq  unemp
0     0.019279  0.028406  0.002204    0.0
1     0.026314  0.032961  0.000977    0.0
2     0.026505  0.031556  0.012409    0.0
3     0.022770  0.030624  0.008312    0.0
4     0.023937  0.030531  0.008987    0.0
...        ...       ...       ...    ...
4538  0.000321  0.000000  0.271618    0.0
4539  0.001216  0.000000  0.271641    0.0
4540  0.001953  0.000000  0.266755    0.0
4541  0.004402  0.002312  0.270990    0.0
4542  0.006029  0.003654  0.271771    0.0

[4543 rows x 4 columns]

In [73]:
print('\033[1mAverage Call Option Prices excluding last 21 days')
indices[['call_5y','call_tips','call_nasdaq','call_unemp']][:-21].mean()

Average Call Option Prices excluding last 21 days


call_5y        0.006549
call_tips      0.006012
call_nasdaq    0.005381
call_unemp     0.013221
dtype: float64

## For value at expiry, we do not have the values for last 21 days

In [74]:
print('\033[1mAverage Discounted Value at Expiry')
disc_values_at_exp.mean()

Average Discounted Value at Expiry


5Y        0.007443
tips      0.005997
nasdaq    0.024952
unemp     0.034829
dtype: float64

## The Values are quite different except for TIPS and treasury compared to the average call option prices. The difference in the values can be explained by the drawbacks of Black-Scholes model:

##  1. The model assumes constant Volatility of the underlying which is not the case in the real world.
##  2. The model assumes the normality of the returns distribution but the real world returns are often distributed with fatter tails.
## So these assumptions of BSM might not depict the reality which is causing the discrepancy between Average Discounted Value at Expiry and Average Call Option Price which are supposed to be similar if the pricing is done right.

## Q9

## To modify the BSM to improve the match between average value at expiry and average call option prices, I think using power law distribution with different parameters for each type of security instead of normal distribution would better describe the way returns are distributed and will improve the match.
## This makes sense economically, since we are not assuming all the financial data to be normal and most of the financial data fits this distribution and hence gives us better results.

In [75]:
print('\033[1mAverage Discounted Value at Expiry')
disc_values_at_exp.mean()

Average Discounted Value at Expiry


5Y        0.007443
tips      0.005997
nasdaq    0.024952
unemp     0.034829
dtype: float64

In [76]:
from scipy.stats import powerlaw

In [77]:
def bsm_call_nasdaq(sigma, S, K, r, t): #Black Scholes Formula
    d1 = (log(S/K) + (r + sigma**2/2)* t)/(sigma * sqrt(t))
    d2 = d1- sigma*sqrt(t)
    c = powerlaw.cdf(d1,0.725) *S - powerlaw.cdf(d2,0.725) * K * math.exp(-r * t)
    return c

In [89]:
call_nasdaq = []
for index, row in indices.iterrows():
    nasdaq = bsm_call_nasdaq(row['nasdaq_std'],row['nasdaq'],row['nasdaq'],row['1M'],21)
    call_nasdaq.append(nasdaq)
print('\033[1mAverage Option Price of Nasdaq Call using powerlaw distribution with parameter 0.725 excluding last 21 days')
st.mean(call_nasdaq[:-21])    

Average Option Price of Nasdaq Call using powerlaw distribution with parameter 0.725 excluding last 21 days


0.023244367643561855

In [79]:
def bsm_call_unemp(sigma, S, K, r, t): #Black Scholes Formula
    d1 = (log(S/K) + (r + sigma**2/2)* t)/(sigma * sqrt(t))
    d2 = d1- sigma*sqrt(t)
    c = powerlaw.cdf(d1,0.9) *S - powerlaw.cdf(d2,0.9) * K * math.exp(-r * t)
    return c

In [88]:
call_unemp = []
for index, row in indices.iterrows():
    unemp = bsm_call_unemp(row['unemp_std'],row['unemp'],row['unemp'],row['1M'],21)
    call_unemp.append(unemp)
print('\033[1mAverage Option Price of Unemployment Call using powerlaw distribution with parameter 0.9 excluding last 21 days')
st.mean(call_unemp[:-21])    

Average Option Price of Unemployment Call using powerlaw distribution with parameter 0.9 excluding last 21 days


0.035057008358676464

In [81]:
def bsm_call_5y(sigma, S, K, r, t): #Black Scholes Formula
    d1 = (log(S/K) + (r + sigma**2/2)* t)/(sigma * sqrt(t))
    d2 = d1- sigma*sqrt(t)
    c = powerlaw.cdf(d1,1.25) *S - powerlaw.cdf(d2,1.25) * K * math.exp(-r * t)
    return c

In [86]:
call_5y = []
for index, row in indices.iterrows():
    val = bsm_call_5y(row['5y_std'],row['5Y'],row['5Y'],row['1M'],21)
    call_5y.append(val)
print('\033[1mAverage Option Price of Treasury Call using powerlaw distribution with parameter 1.25 excluding last 21 days')
st.mean(call_5y[:-21])    

Average Option Price of Treasury Call using powerlaw distribution with parameter 1.25 excluding last 21 days


0.00737024262566779

In [83]:
def bsm_call_tips(sigma, S, K, r, t): #Black Scholes Formula
    d1 = (log(S/K) + (r + sigma**2/2)* t)/(sigma * sqrt(t))
    d2 = d1- sigma*sqrt(t)
    c = powerlaw.cdf(d1,1.3) *S - powerlaw.cdf(d2,1.3) * K * math.exp(-r * t)
    return c

In [87]:
call_tips = []
for index, row in indices.iterrows():
    val = bsm_call_tips(row['tips_std'],row['tips'],row['tips'],row['1M'],21)
    call_tips.append(val)
print('\033[1mAverage Option Price of TIPS Call using powerlaw distribution with parameter 1.3 excluding last 21 days')
st.mean(call_tips[:-21]) 

Average Option Price of TIPS Call using powerlaw distribution with parameter 1.3 excluding last 21 days


0.0060200874003258084

## Usually the financial data fits te power law distribution with parameter in the range of 0 to 3, so I used the parameters in those range to improve the match and included those which had the best match. 
## We can see that using Powerlaw distributions for underlying returns with different parameters, the BSM is producing better prices on average than the normal distribution. And the average option prices are almost equal to average value at expiry. 
## Since most of the financial data can be fit using powerlaw distributions, using those would be appropriate economically.